### ISSUE TITLE
Possible false positive "Undocumented API"

### ISSUE THREAD

**Issue by wichtounet** [SonarOpenCommunity/sonar-cxx#751]
Hi, 

I have found something that troubles me, but perhaps it is expected behaviour. 

In my code, I have a macro that is used to generate functors and then I generate several functors:

```cpp
#define MUL_FUNCTOR(name, ...)                        \
    struct name {                                     \
        template <typename A, typename B, typename C> \
        static void apply(A&& a, B&& b, C& c) {       \
            __VA_ARGS__;                              \
        }                                             \
    };

MUL_FUNCTOR(default_gemm, c = etl::mul(a, b))
MUL_FUNCTOR(lazy_gemm, c = etl::lazy_mul(a, b))
//...
```

Every line with MUL_FUNCTOR call is annotated with "Undocumented API: xxx". 

In a sense, it is true since it a struct and should be annotated but then it is a macro expansion and should perhaps not be documented.  

Is that expected ? 

Thanks

**Comment by guwirth** [comment]
The preprocessor is extracting the macros, the c++ parser is creating the AST. The documentation check is on top of the AST. Not sure if it is possible to detect that code is original from a macro?

Other thing is that from my point of view `MUL_FUNCTOR` should be documented because it's a "*function*" in the global namespace.

**Comment by lcintrat** [comment]
Hi,

From your snippet, 4 issues are expected:
 - `struct default_gemm`
 - `member default_gemm::apply`
 - `struct lazy_gemm`
 - `member lazy_gemm::apply`

I agree with @guwirth for the parsing, the visitor visits the AST, and retrieving the original tokens would be difficult. Same thing for the `MUL_FUNCTOR` documentation. However, Doxygen seems to handle macro documentation...

**Comment by wichtounet** [comment]
The MUL_FUNCTOR definition has no issue on it. I agree that it should be documented and that in this case, an issue should probably be created. 

I checked and there is indeed two issues for each MUL_FUNCTOR expansion, so that is coherent at least. 

I also checked, by adding comments inside the macro, it removes the issues. I think this is the best way to handle it, but isn't it contrary to the standard which states that the removal of comments should be done before preprocessing ?

**Comment by wichtounet** [comment]
I updated my last message, I was not testing it correctly :(

How would you go about fixing these issues ? Simply setting them as false positives in Sonar ?

Adding comments inside the macro should not work since comments are replaced by spaces during tokenization which is before macro expansion.

**Comment by guwirth** [comment]
@wichtounet please reopen if still an issue.
